README
- sIBL_GUI For Maya - Helper Script -

- Installation -
Extract "sIBL_GUI For Maya.zip" content into your Maya Preferences folder, for example on Windows: "C:\Documents and Settings\$USER\Mes documents\maya\2011\". Start Maya, and you should find a new "sIBL_GUI" shelf.
You can add the following lines into your "userSetup.mel" if you want the Socket Port to be opened automatically on Maya startup:

	commandPort -n ( "127.0.0.1:"  + `optionVar -q "sIBL_GUI_commandPort"` );

- Usage -

   -  "p" icon launches "sIBL_GUI For Maya Preferences" window where you define "sIBL_GUI" executable path, Loader Script path and Remote Connection Port.
   -  "O" icon launches "sIBL_GUI".
   -  "e" icon executes the "Loader Script".
   -  "e" icon launches the cleanup function.

- About -
sIBL_GUI by Thomas Mansencal - 2008 - 2012
Copyright (C) 2008 - 2012 - Thomas Mansencal - thomas.mansencal@gmail.com
This software is released under terms of GNU GPL V3 license: http://www.gnu.org/licenses/
http://www.thomasmansencal.com/

CHANGELOG
sIBL_GUI For Maya - 2.0.1
=========================

Change Log
==========

2.0.1
-----

-  Fixed executable path filtering rule on Mac Os X.

2.0.0
-----

-  Code refactor for sIBL_GUI 4.0.0.
-  Dropped support for Maya versions prior Maya 2011.

1.1.2
-----

-  Added Smart Ibl nodes removal shelf button.

1.1.1
-----

-  Updated code to directly support "sIBL_GUI.app" as sIBL_GUI executable path in Mac Os X.

1.1.0
-----

-  Repackaged with modified header.

1.0.9
-----

-  Added online links buttons.
-  Added sIBL_GUI Logo for Qt helper script version.
-  Ui updated.

1.0.8
-----

-  Added Maya 2011 support.
-  Refactored the code into a "sIBL_GUI_For_Maya" module.
-  Updated the icons.

1.0.7
-----

-  Fixed a missing Procedure Call.

1.0.6
-----

-  Command Port Option Var is now correctly stored when you open the Command Port.

1.0.5
-----

-  Modified Command Port opening for Windows Vista x64 and Windows 7 x64 support.

1.0.4
-----

-  Corrected a problem related to "platform.system()" command on Windows.

1.0.3
-----

-  Compatibility with Mac Os X and Linux.

1.0.2
-----

-  Rolled Back Output Loader Script Name.

1.0.1
-----

-  Changed Output Loader Script Name.

1.0.0
-----

-  Initial Release of sIBL_GUI For Maya.
 
About
-----

| sIBL_GUI by Thomas Mansencal – 2008 - 2012
| Copyright© 2008 - 2012 – Thomas Mansencal – `thomas.mansencal@gmail.com <mailto:thomas.mansencal@gmail.com>`_
| This software is released under terms of GNU GPL V3 license: http://www.gnu.org/licenses/
| `http://www.thomasmansencal.com/ <http://www.thomasmansencal.com/>`_

utilities/gatherFiles.sh
#/bin/bash
echo ----------------------------------------------------------------
echo sIBL_GUI For Maya - Files Gathering
echo ----------------------------------------------------------------

export UTILITIES=/Users/KelSolaar/Documents/Developement/sIBL_GUI/utilities
export PROJECT=/Users/KelSolaar/Documents/Developement/sIBL_GUI_For_Maya

export SOURCE=$PROJECT/src/
export RELEASES=$PROJECT/releases/
export BUILD=$RELEASES/build
export REPOSITORY=$RELEASES/repository

#! Gathering folder cleanup.
echo ----------------------------------------------------------------
echo Gathering Folder Cleanup - Begin
echo ----------------------------------------------------------------
rm -rf $BUILD
rm -rf $REPOSITORY/*
echo ----------------------------------------------------------------
echo Gathering Folder Cleanup - End
echo ----------------------------------------------------------------

#! Extra files cleanup.
echo ----------------------------------------------------------------
echo Extra Files Cleanup - Begin
echo ----------------------------------------------------------------
python $UTILITIES/recursiveRemove.py ./ .DS_Store
echo ----------------------------------------------------------------
echo Extra Files Cleanup - End
echo ----------------------------------------------------------------

#! Change log gathering.
echo ----------------------------------------------------------------
echo Change Log Gathering - Begin
echo ----------------------------------------------------------------
cp $RELEASES/Change_Log.html $REPOSITORY/
echo ----------------------------------------------------------------
echo Change Log Gathering - End
echo ----------------------------------------------------------------

echo ----------------------------------------------------------------
echo Files Gathering - Begin
echo ----------------------------------------------------------------
mkdir -p $BUILD/prefs/icons
mkdir -p $BUILD/prefs/shelves
mkdir $BUILD/scripts
cp $PROJECT/README $BUILD/
cp $SOURCE/prefs/icons/*.png $BUILD/prefs/icons/
cp $SOURCE/prefs/shelves/shelf_sIBL_GUI.mel $BUILD/prefs/shelves/shelf_sIBL_GUI.mel
cp $SOURCE/scripts/sIBL_GUI_For_Maya.py $BUILD/scripts/sIBL_GUI_For_Maya.py
cd $BUILD/
zip -r $REPOSITORY/sIBL_GUI_For_Maya.zip .
echo ----------------------------------------------------------------
echo Files Gathering - End
echo ----------------------------------------------------------------

src/scripts/sIBL_GUI_For_Maya_Qt.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
**sIBL_GUI_For_Maya.py

**Platform:**
	Windows, Linux, Mac Os X.

**Description:**
	Maya sIBL_GUI Helper Script For Maya 2011 And Higher.

**Others:**

"""

#**********************************************************************************************************************
#***	External imports.
#**********************************************************************************************************************
import maya.cmds as cmds
import maya.mel as mel
import os
import platform
import re

#**********************************************************************************************************************
#***	Module attributes.
#**********************************************************************************************************************
__author__ = "Thomas Mansencal"
__copyright__ = "Copyright (C) 2008 - 2011 - Thomas Mansencal"
__license__ = "GPL V3.0 - http://www.gnu.org/licenses/"
__maintainer__ = "Thomas Mansencal"
__email__ = "thomas.mansencal@gmail.com"
__status__ = "Production"

LOADER_SCRIPTS_DIRECTORY = "HDRLabs/sIBL_GUI/io/loaderScripts/"
LOADER_SCRIPT = "sIBL_Maya_Import.mel"
HDRLABS_URL = "http://www.hdrlabs.com"
WINDOWS_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Windows"
DARWIN_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Darwin"
LINUX_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Linux"
APPLICATION_THREAD_URL = "http://www.hdrlabs.com/cgi-bin/forum/YaBB.pl?num=1271609371"

#**********************************************************************************************************************
#***	Module classes and definitions.
#**********************************************************************************************************************
class Environment(object):
	"""
	This Class Provides Methods To Manipulate Environment Variables.
	"""

	def __init__(self, variable = None):
		"""
		This Method Initializes The Class.

		:param variable: Variable To Manipulate. (String)
		"""

		# --- Setting Class Attributes. ---
		self._variable = variable

	def getPath(self):
		"""
		This Method Gets The Chosen Environment Variable Path As A String.

		:return: Variable Path. (String)
		"""

		if self._variable:
			for param in os.environ:
				if(self._variable == param): return os.environ[param]

def getSystemApplicationDataDirectory():
	"""
	This Definition Gets The System Application Data Directory.

	:return: User Application Data Directory. (String)
	"""

	if platform.system() == "Windows" or platform.system() == "Microsoft":
		environmentVariable = Environment("APPDATA")
		return environmentVariable.getPath()

	elif platform.system() == "Darwin":
		environmentVariable = Environment("HOME")
		return os.path.join(environmentVariable.getPath(), "Library/Preferences")

	elif platform.system() == "Linux":
		environmentVariable = Environment("HOME")
		return environmentVariable.getPath()

def storeCommandPortOptionVar():
	"""
	This Definition Stores The Command Port In An Option Var.
	"""

	cmds.optionVar(iv = ("sIBL_GUI_Command_Port", cmds.intField("sIBL_CommandPort_IntField", query = True, value = True)))

def storeExecutablePathOptionVar():
	"""
	This Definition Stores sIBL_GUI Executable Path In An Option Var.
	"""

	cmds.optionVar(sv = ("sIBL_GUI_Executable_Path", cmds.textField("sIBL_ExecutablePath_TextField", query = True, text = True)))

def openCommandPort():
	"""
	This Definition Opens The Command Port.
	"""

	try:
		cmds.commandPort(name = "127.0.0.1:" + str(cmds.intField("sIBL_CommandPort_IntField", query = True, value = True)))
	except:
		mel.eval("warning(\"sIBL_GUI | Command Port Is Already Open Or Can't Be Opened!\");")

	storeCommandPortOptionVar()

def executableFileBrowser():
	"""
	This Definition Provides A Browser.
	"""

	fileName = cmds.fileDialog2(ds = 2, fileFilter = "All Files (*.*)", fm = (not platform.system() == "Darwin" and 1 or 3))
	fileName = fileName and fileName[0] or None

	if fileName:
		if fileName.endswith("sIBL_GUI.exe") or fileName.endswith("sIBL_GUI.app") or fileName.endswith("sIBL_GUI"):
			cmds.textField("sIBL_ExecutablePath_TextField", edit = True, text = fileName)
			storeExecutablePathOptionVar()
		else:
			mel.eval("warning(\"sIBL_GUI | Chosen Executable Path Is Invalid!\");")

def sIBL_CommandPortIntSlider_OnEdit():
	"""
	This Definition Is Triggered By sIBL_CommandPortIntSlider Edit.
	"""

	cmds.intField("sIBL_CommandPort_IntField", edit = True, value = cmds.intSlider("sIBL_CommandPort_IntSlider", query = True, value = True))

	storeCommandPortOptionVar()

def sIBL_CommandPortIntField_OnEdit():
	"""
	This Definition Is Triggered By sIBL_CommandPortIntField Edit.
	"""

	cmds.intSlider("sIBL_CommandPort_IntSlider", edit = True, value = cmds.intField("sIBL_CommandPort_IntField", query = True, value = True))

	storeCommandPortOptionVar()

def sIBL_ExecutablePathTextField_OnEdit():
	"""
	This Definition Is Triggered By sIBL_ExecutablePathTextField Edit.
	"""

	textFieldContent = cmds.textField("sIBL_ExecutablePath_TextField", query = True, text = True)
	if textFieldContent.endswith("sIBL_GUI.exe") or textFieldContent.endswith("sIBL_GUI.app") or  textFieldContent.endswith("sIBL_GUI"):
		storeExecutablePathOptionVar()
	else:
		mel.eval("warning(\"sIBL_GUI | Chosen Executable Path Is Invalid!\");")

def openUrl(url):
	"""
	This Definition Opens HDRLabs Thread.

	:param url: Url To Open. (String)
	"""
	cmds.launch(web = url)

def getApplication_Button_OnClicked():
	"""
	This Definition Opens Online Repository.
	"""

	if platform.system() == "Windows" or platform.system() == "Microsoft":
		url = WINDOWS_RELEASE_URL
	elif platform.system() == "Darwin":
		url = DARWIN_RELEASE_URL
	elif platform.system() == "Linux":
		url = LINUX_RELEASE_URL

	openUrl(url)

def hdrlabs_Button_OnClicked():
	"""
	This Definition Opens HDRLabs Thread.
	"""

	openUrl(HDRLABS_URL)

def applicationThread_Button_OnClicked():
	"""
	This Definition Opens sIBL_GUI Thread.
	"""

	openUrl(APPLICATION_THREAD_URL)

def openPreferences():
	"""
	This Definition Launchs sIBL_GUI For Maya Preferences.
	"""

	cmds.windowPref(enableAll = False)

	if (cmds.window("sIBL_GUI_For_Maya_Windows", exists = True)):
		cmds.deleteUI("sIBL_GUI_For_Maya_Windows")

	cmds.window("sIBL_GUI_For_Maya_Windows",
		title = "sIBL_GUI For Maya - Preferences",
		width = 8,
		height = 8,
		rtf = True,
		toolbox = True)

	horizontalSpacing = 8
	verticalSpacing = 16
	separatorHeight = 2

	cmds.columnLayout(adjustableColumn = True, rowSpacing = verticalSpacing, columnOffset = ["both", horizontalSpacing ])

	cmds.separator(height = separatorHeight, style = "none")

	cmds.picture(image = "sIBL_GUI_Small_Logo.png")

	cmds.frameLayout(label = "sIBL_GUI Executable Path", cll = False, li = 4, borderStyle = "etchedOut", mh = 4, mw = 4)
	cmds.rowLayout("executablePathRowLayout", numberOfColumns = 2, adjustableColumn = 1, columnAlign2 = ["center", "center"], columnAttach = [(1, "both", horizontalSpacing), (2, "both", horizontalSpacing)])
	cmds.textField("sIBL_ExecutablePath_TextField", cc = "sIBL_GUI_For_Maya.sIBL_ExecutablePathTextField_OnEdit()")
	cmds.button("fileBrowser_Button", label = "...", al = "center", command = "sIBL_GUI_For_Maya.executableFileBrowser()")
	cmds.setParent(upLevel = True)
	cmds.setParent(upLevel = True)

	sIBL_GUI_Executable_Path = cmds.optionVar(q = "sIBL_GUI_Executable_Path")
	if sIBL_GUI_Executable_Path != 0:
		cmds.textField("sIBL_ExecutablePath_TextField", edit = True, text = sIBL_GUI_Executable_Path)

	cmds.frameLayout(label = "sIBL_GUI Command Port", cll = False, li = 4, borderStyle = "etchedOut", mh = 4, mw = 4)
	cmds.rowLayout("commandPortRowLayout", numberOfColumns = 3, adjustableColumn = 2, columnAlign3 = ["center", "center", "center"], columnAttach = [(1, "both", horizontalSpacing), (2, "both", horizontalSpacing), (3, "both", horizontalSpacing)])
	cmds.intField("sIBL_CommandPort_IntField", minValue = 0, maxValue = 65535, value = 2048, cc = "sIBL_GUI_For_Maya.sIBL_CommandPortIntField_OnEdit()")
	cmds.intSlider("sIBL_CommandPort_IntSlider", min = 0, max = 65535 , value = cmds.intField("sIBL_CommandPort_IntField", query = True, value = True) , step = 1, cc = "sIBL_GUI_For_Maya.sIBL_CommandPortIntSlider_OnEdit()")
	cmds.button("commandPort_Button", label = "Open Port", al = "center", command = "sIBL_GUI_For_Maya.openCommandPort()")
	cmds.setParent(upLevel = True)
	cmds.setParent(upLevel = True)

	cmds.frameLayout(label = "Online", cll = False, li = 4, borderStyle = "etchedOut", mh = 4, mw = 4)
	cmds.rowLayout("onlineRowLayout", numberOfColumns = 3, adjustableColumn = 3, columnAlign3 = ["center", "center", "center"], columnAttach = [(1, "both", horizontalSpacing), (2, "both", horizontalSpacing), (3, "both", horizontalSpacing)])
	cmds.button("getApplication_Button", label = "Get sIBL_GUI", al = "center", command = "sIBL_GUI_For_Maya.getApplication_Button_OnClicked()")
	cmds.button("hdrlabs_Button", label = "Visit HDRLabs", al = "center", command = "sIBL_GUI_For_Maya.hdrlabs_Button_OnClicked()")
	cmds.button("applicationThread_Button", label = "Visit sIBL_GUI Thread", al = "right", command = "sIBL_GUI_For_Maya.applicationThread_Button_OnClicked()")
	cmds.setParent(upLevel = True)
	cmds.setParent(upLevel = True)

	cmds.separator(height = separatorHeight, style = "none")

	sIBL_GUI_Command_Port = int(cmds.optionVar(q = "sIBL_GUI_Command_Port"))
	if sIBL_GUI_Command_Port != 0:
		cmds.intField("sIBL_CommandPort_IntField", edit = True, value = sIBL_GUI_Command_Port)

	cmds.showWindow("sIBL_GUI_For_Maya_Windows")
	cmds.windowPref(enableAll = True)

def launchApplication():
	"""
	This Definition Launchs sIBL_GUI.
	"""

	sIBL_GUI_Executable_Path = cmds.optionVar(q = "sIBL_GUI_Executable_Path")
	if sIBL_GUI_Executable_Path != 0:
		if platform.system() == "Windows" or platform.system() == "Microsoft":
			os.system("start /D" + "\"" + os.path.dirname(sIBL_GUI_Executable_Path) + "\"" + " " + sIBL_GUI_Executable_Path.replace(" ", "\" \""))
		elif platform.system() == "Darwin":
			os.system("open -a " + sIBL_GUI_Executable_Path)
		elif platform.system() == "Linux":
			os.system("\"" + sIBL_GUI_Executable_Path + "\" &")
	else:
		mel.eval("warning(\"sIBL_GUI | No sIBL_GUI Executable Path Defined!\");")

def executeLoaderScript():
	"""
	This Definition Executes sIBL_GUI Maya Loader Script.
	"""

	systemApplicationDataDirectory = getSystemApplicationDataDirectory()

	if systemApplicationDataDirectory:
		loaderScript = os.path.normpath(os.path.join(systemApplicationDataDirectory, LOADER_SCRIPTS_DIRECTORY, LOADER_SCRIPT))

		if platform.system() == "Windows" or platform.system() == "Microsoft":
			loaderScript = loaderScript.replace("\\", "\\\\")

		if os.path.exists(loaderScript):
			mel.eval("source \"" + loaderScript + "\"")
		else:
			mel.eval("warning(\"sIBL_GUI | No Maya Loader Script Found!\");")

def deleteSmartIblNodes():
	"""
	This Definition Deletes Smart Ibl And Lightsmith Lights Nodes.
	"""

	nodes = []
	selection = cmds.ls(sl = True, l = True)
	if selection:
		prefixes = []
		for node in selection:
			relatives = [node]
			relatives.extend(cmds.listRelatives(node, f = True, ad = True))
			for relative in relatives:
				if re.search(r"\w*_Root$", relative):
					if relative.replace("_Root", "_Support").split("|")[-1] in cmds.listRelatives(relative, ad = True):
						prefixes.append(relative.replace("_Root", "").split("|")[-1])
		for prefix in prefixes:
			userChoice = cmds.confirmDialog(title="sIBL_GUI", message="Nodes With Following Prefix : '%s' Are Planned For Deletion! Would You Like To Proceed?" % prefix, button=["Yes","No"], defaultButton="Yes", cancelButton="No", dismissString="No")
			if userChoice == "Yes":
				nodes.extend(sorted(cmds.ls(prefix + "*", l = True)))
	else:
		userChoice = cmds.confirmDialog(title="sIBL_GUI", message="Smart Ibl Nodes Are Planned For Deletion! Would You Like To Proceed?", button=["Yes","No"], defaultButton="Yes", cancelButton="No", dismissString="No")
		if userChoice == "Yes":
			nodes.extend(sorted(cmds.ls("sIBL*", l = True)))

	for node in nodes:
		if cmds.objExists(node):
			print("sIBL_GUI | Deleting Node: '%s'!" % node)
			cmds.delete(node)

#**********************************************************************************************************************
#***	Python End.
#**********************************************************************************************************************

src/scripts/sIBL_GUI_For_Maya.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
**sIBL_GUI_For_Maya.py

**Platform:**
	Windows, Linux, Mac Os X.

**Description:**
	Maya sIBL_GUI Helper Script for Maya 2011 and higher.

**Others:**

"""

#**********************************************************************************************************************
#***	External imports.
#**********************************************************************************************************************
import maya.cmds as cmds
import maya.mel as mel
import os
import platform
import re
import sys

#**********************************************************************************************************************
#***	Module attributes.
#**********************************************************************************************************************
__author__ = "Thomas Mansencal"
__copyright__ = "Copyright (C) 2008 - 2012 - Thomas Mansencal"
__license__ = "GPL V3.0 - http://www.gnu.org/licenses/"
__maintainer__ = "Thomas Mansencal"
__email__ = "thomas.mansencal@gmail.com"
__status__ = "Production"

__all__ = ["HDRLABS_URL",
		"WINDOWS_RELEASE_URL",
		"DARWIN_RELEASE_URL",
		"LINUX_RELEASE_URL",
		"APPLICATION_THREAD_URL",
		"openPreferences",
		"launchApplication",
		"executeLoaderScript",
		"deleteSmartIblNodes"]

HDRLABS_URL = "http://www.hdrlabs.com"
WINDOWS_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Windows"
DARWIN_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Darwin"
LINUX_RELEASE_URL = "http://kelsolaar.hdrlabs.com/?dir=./sIBL_GUI/Repository/Builds/Linux"
APPLICATION_THREAD_URL = "http://www.hdrlabs.com/cgi-bin/forum/YaBB.pl?num=1271609371"

#**********************************************************************************************************************
#***	Module classes and definitions.
#**********************************************************************************************************************
class Environment(object):
	"""
	This class provides methods to manipulate environment variables.
	"""

	def __init__(self, variable=None):
		"""
		This method initializes the class.

		:param variable: Variable to manipulate. ( String )
		"""

		self._variable = variable

	def getValue(self):
		"""
		This methods returns given environment variable value.

		:return: Variable value. ( String )
		"""

		if self._variable:
			return os.environ.get(self._variable)

def _setExecutablePathOptionVar():
	"""
	This definition sets **executablePath** optionVar.
	"""

	_setOptionVar("sIBL_GUI_executablePath", cmds.textField("Executable_Path_textField", query=True, text=True))

def _setLoaderScriptPathOptionVar():
	"""
	This definition sets **sIBL_GUI_loaderScriptPath** optionVar.
	"""

	_setOptionVar("sIBL_GUI_loaderScriptPath", cmds.textField("Loader_Script_Path_textField", query=True, text=True))

def _setCommandPortOptionVar():
	"""
	This definition sets **sIBL_GUI_commandPort** optionVar.
	"""

	_setOptionVar("sIBL_GUI_commandPort", cmds.intSliderGrp("Command_Port_intSliderGrp", query=True, value=True))

def _setOptionVar(name, value):
	"""
	This definition stores given optionVar with given value.
	
	:param name: OptionVar name. ( String )
	:param value: OptionVar value. ( Object )
	"""

	cmds.optionVar(sv=(name, value))

def _openUrl(url):
	"""
	This definition opens given url.

	:param url: Url to open. ( String )
	"""

	cmds.launch(web=url)

def _Executable_Path_button__command(state=None):
	"""
	This definition is triggered by **Executable_Path_button** widget.

	:param state: Button state. ( Boolean )
	"""

	fileName = cmds.fileDialog2(ds=2, fileFilter="All Files (*.*)", fm=(not platform.system() == "Darwin" and 1 or 3))
	fileName = fileName and fileName[0] or None
	if not fileName:
		return

	if fileName.endswith("sIBL_GUI.exe") or re.search("sIBL_GUI[\d ]*\.app", fileName) or fileName.endswith("sIBL_GUI"):
		cmds.textField("Executable_Path_textField", edit=True, text=fileName)
		_setExecutablePathOptionVar()
	else:
		mel.eval("warning(\"sIBL_GUI | Chosen executable path is invalid!\");")

def _Executable_Path_textField__changeCommand(value):
	"""
	This definition is triggered by **Executable_Path_textField** widget.

	:param value: Value. ( String )
	"""

	if os.path.exists(value) and \
		(value.endswith("sIBL_GUI.exe") or \
		value.endswith("sIBL_GUI.app") or \
		value.endswith("sIBL_GUI")):
		_setExecutablePathOptionVar()
	else:
		mel.eval("warning(\"sIBL_GUI | Chosen executable path is invalid!\");")

def _Loader_Script_Path_button__command(state=None):
	"""
	This definition is triggered by **Loader_Script_Path_button** widget.

	:param state: Button state. ( Boolean )
	"""

	fileName = cmds.fileDialog2(ds=2, fileFilter="All Files (*.*)", fm=4)
	fileName = fileName and fileName[0] or None
	if not fileName:
		return

	cmds.textField("Loader_Script_Path_textField", edit=True, text=fileName)
	_setLoaderScriptPathOptionVar()

def _Loader_Script_Path_textField__changeCommand(value):
	"""
	This definition is triggered by **_Loader_Script_Path_textField** widget.

	:param value: Value. ( String )
	"""

	if os.path.exists(value):
		_setLoaderScriptPathOptionVar()
	else:
		mel.eval("warning(\"sIBL_GUI | Chosen Loader Script path is invalid!\");")

def _Command_Port_intSliderGrp__changeCommand(value):
	"""
	This definition is triggered by **Command_Port_intSliderGrp** widget.

	:param value: Value. ( Float )
	"""

	_setCommandPortOptionVar()

def _Command_Port_button__command(state=None):
	"""
	This definition opens the command port.

	:param state: Button state. ( Boolean )
	"""

	try:
		cmds.commandPort(name="127.0.0.1:" + str(cmds.intSliderGrp("Command_Port_intSliderGrp", query=True, value=True)))
	except:
		mel.eval("warning(\"sIBL_GUI | Command port is already open or can't be opened!\");")

	_setCommandPortOptionVar()

def _Get_Application_button__command(state=None):
	"""
	This definition opens Online Repository.

	:param state: Button state. ( Boolean )
	"""

	if platform.system() == "Windows" or platform.system() == "Microsoft":
		url = WINDOWS_RELEASE_URL
	elif platform.system() == "Darwin":
		url = DARWIN_RELEASE_URL
	elif platform.system() == "Linux":
		url = LINUX_RELEASE_URL

	_openUrl(url)

def _HDRlabs_button__command(state=None):
	"""
	This definition opens HDRLabs thread.

	:param state: Button state. ( Boolean )
	"""

	_openUrl(HDRLABS_URL)

def _Application_Thread_button__command(state=None):
	"""
	This definition opens sIBL_GUI thread.

	:param state: Button state. ( Boolean )
	"""

	_openUrl(APPLICATION_THREAD_URL)

def _sIBL_GUI_For_Maya_window():
	"""
	This definition launches **sIBL_GUI For Maya Preferences** window.
	"""

	cmds.windowPref(enableAll=False)

	if (cmds.window("_sIBL_GUI_For_Maya_window", exists=True)):
		cmds.deleteUI("_sIBL_GUI_For_Maya_window")

	cmds.window("_sIBL_GUI_For_Maya_window",
		title="sIBL_GUI For Maya - Preferences",
		sizeable=False)

	horizontalSpacing = 8

	cmds.columnLayout(adjustableColumn=True)

	cmds.picture(image="sIBL_GUI_Small_Logo.png")

	cmds.frameLayout(label="sIBL_GUI Executable Path", cll=False, li=4, borderStyle="etchedOut", mh=4, mw=4)
	cmds.rowLayout(numberOfColumns=2, adjustableColumn=1, columnAlign2=["center", "center"], columnAttach=[(1, "both", horizontalSpacing), (2, "both", horizontalSpacing)])
	cmds.textField("Executable_Path_textField", cc=_Executable_Path_textField__changeCommand)
	cmds.button("Executable_Path_button", label="...", al="center", command=_Executable_Path_button__command)
	cmds.setParent(upLevel=True)
	cmds.setParent(upLevel=True)

	executablePath = cmds.optionVar(q="sIBL_GUI_executablePath")
	if executablePath:
		cmds.textField("Executable_Path_textField", edit=True, text=executablePath)

	cmds.frameLayout(label="Loader Script Path", cll=False, li=4, borderStyle="etchedOut", mh=4, mw=4)
	cmds.rowLayout(numberOfColumns=2, adjustableColumn=1, columnAlign2=["center", "center"], columnAttach=[(1, "both", horizontalSpacing), (2, "both", horizontalSpacing)])
	cmds.textField("Loader_Script_Path_textField", cc=_Loader_Script_Path_textField__changeCommand)
	cmds.button("Loader_Script_Path_button", label="...", al="center", command=_Loader_Script_Path_button__command)
	cmds.setParent(upLevel=True)
	cmds.setParent(upLevel=True)

	sIBL_GUI_loaderScriptPath = cmds.optionVar(q="sIBL_GUI_loaderScriptPath")
	if sIBL_GUI_loaderScriptPath:
		cmds.textField("Loader_Script_Path_textField", edit=True, text=sIBL_GUI_loaderScriptPath)

	cmds.frameLayout(label="Command Port", cll=False, li=4, borderStyle="etchedOut", mh=4, mw=4)
	cmds.rowLayout(numberOfColumns=2, adjustableColumn=1, columnAlign2=["center", "center"], columnAttach=[(1, "both", horizontalSpacing), (2, "both", horizontalSpacing)])
	cmds.intSliderGrp("Command_Port_intSliderGrp", field=True, minValue=0, maxValue=65535, value=2048, cc=_Command_Port_intSliderGrp__changeCommand)
	cmds.button("Command_Port_button", label="Open Port", al="center", command=_Command_Port_button__command)
	cmds.setParent(upLevel=True)
	cmds.setParent(upLevel=True)

	sIBL_GUI_commandPort = int(cmds.optionVar(q="sIBL_GUI_commandPort"))
	if sIBL_GUI_commandPort:
		cmds.intSliderGrp("Command_Port_intSliderGrp", edit=True, value=sIBL_GUI_commandPort)

	cmds.frameLayout(label="Online", cll=False, li=4, borderStyle="etchedOut", mh=4, mw=4)
	cmds.rowLayout(numberOfColumns=3, adjustableColumn=3, columnAlign3=["center", "center", "center"], columnAttach=[(1, "both", horizontalSpacing), (2, "both", horizontalSpacing), (3, "both", horizontalSpacing)])
	cmds.button("Get_Application_button", label="Get sIBL_GUI ...", al="center", command=_Get_Application_button__command)
	cmds.button("HDRlabs_button", label="Visit HDRLabs ...", al="center", command=_HDRlabs_button__command)
	cmds.button("Application_Thread_button", label="Visit sIBL_GUI Thread ...", al="right", command=_Application_Thread_button__command)
	cmds.setParent(upLevel=True)
	cmds.setParent(upLevel=True)

	cmds.showWindow("_sIBL_GUI_For_Maya_window")
	cmds.windowPref(enableAll=True)

def openPreferences():
	"""
	This definition launches **sIBL_GUI For Maya - Preferences** window.
	"""

	_sIBL_GUI_For_Maya_window()
	return True

def launchApplication():
	"""
	This definition launches **sIBL_GUI**.

	:return: Definition sucess. ( Boolean )
	"""

	executablePath = cmds.optionVar(q="sIBL_GUI_executablePath")
	if executablePath:
		if platform.system() == "Windows" or platform.system() == "Microsoft":
			os.system("start /D" + "\"" + os.path.dirname(executablePath) + "\"" + " " + executablePath.replace(" ", "\" \""))
		elif platform.system() == "Darwin":
			os.system("open -a " + executablePath.replace(" ", "\ "))
		elif platform.system() == "Linux":
			os.system("\"" + executablePath.replace(" ", "\ ") + "\" &")
		return True
	else:
		mel.eval("warning(\"sIBL_GUI | No sIBL_GUI executable path defined!\");")
		cmds.confirmDialog(title="sIBL_GUI | Warning", message="No sIBL_GUI executable path defined!\nPlease define one in preferences!", button=["Ok"], defaultButton="Ok")
		openPreferences()

def executeLoaderScript():
	"""
	This definition executes **sIBL_GUI** Loader Script.

	:return: Definition sucess. ( Boolean )
	"""

	loaderScriptPath = cmds.optionVar(q="sIBL_GUI_loaderScriptPath")
	if loaderScriptPath:
		if os.path.exists(loaderScriptPath):
			if re.search(r"\.mel$", loaderScriptPath):
				mel.eval("source \"" + loaderScriptPath + "\"")
			elif re.search(r"\.pyc?$", loaderScriptPath):
				path = os.path.dirname(loaderScriptPath)
				path not in sys.path and sys.path.append(path)
				module = os.path.splitext(os.path.basename(loaderScriptPath))[0]
				import_ = __import__(module)
				reload(import_)
				if hasattr(import_, "Setup"):
					setup = import_.Setup()
					setup.execute()
			return True
		else:
			mel.eval("error(\"sIBL_GUI | Loader Script doesn't exists!\");")
	else:
		mel.eval("warning(\"sIBL_GUI | No Loader Script found!\");")
		cmds.confirmDialog(title="sIBL_GUI | Warning", message="No Loader Script found!\nPlease define one in preferences!", button=["Ok"], defaultButton="Ok")
		openPreferences()

def deleteSmartIblNodes():
	"""
	This definition deletes **Smart Ibl** and **Lightsmith** lights nodes.
	
	:return: Definition sucess. ( Boolean )
	"""

	nodes = []
	selection = cmds.ls(sl=True, l=True, dag=True)
	if selection:
		prefixes = []
		for node in selection:
			relatives = [node]
			nodeRelatives = cmds.listRelatives(node, f=True, ad=True)
			nodeRelatives and relatives.extend(nodeRelatives)
			for relative in relatives:
				if re.search(r"\w*_Root$", relative):
					if relative.replace("_Root", "_Support").split("|")[-1] in cmds.listRelatives(relative, ad=True):
						prefixes.append(relative.replace("_Root", "").split("|")[-1])
		for prefix in prefixes:
			userChoice = cmds.confirmDialog(title="sIBL_GUI", message="Nodes with following prefix : '%s' are planned for deletion! Would you like to proceed?" % prefix, button=["Yes", "No"], defaultButton="Yes", cancelButton="No", dismissString="No")
			if userChoice == "Yes":
				nodes.extend(sorted(cmds.ls(prefix + "*", l=True)))
	else:
		result = cmds.promptDialog(title="sIBL_GUI | Nodes Deletion",
	                message="Enter scene 'Smart Ibl' Nodes prefix ( 'sIBL' ):",
	                text="sIBL",
	                button=["Ok", "Cancel"],
	                defaultButton="Ok",
	                cancelButton="Cancel",
	                dismissString="Cancel")

		if result == "Ok":
			prefix = cmds.promptDialog(query=True, text=True)
			nodes.extend(sorted(cmds.ls("%s*" % prefix, l=True)))

	for node in nodes:
		if cmds.objExists(node):
			print("sIBL_GUI | Deleting node: '%s'!" % node)
			cmds.delete(node)

	return True
